test/lit/passes/gufa-tnh.wast

```diff
@@ -0,0 +1,1905 @@
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (ref.cast $B
+  ;; CHECK-NEXT:    (local.get $x)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $called (param $x (ref null $A))
+    (drop
+      (ref.cast $B
+        (local.get $x)
+      )
+    )
+  )
+
+  ;; CHECK:      (func $caller (type $anyref_=>_none) (param $x anyref)
+  ;; CHECK-NEXT:  (call $called
+  ;; CHECK-NEXT:   (ref.cast $B
+  ;; CHECK-NEXT:    (local.get $x)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (call $called
+  ;; CHECK-NEXT:   (ref.cast $B
+  ;; CHECK-NEXT:    (local.get $x)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (call $called
+  ;; CHECK-NEXT:   (ref.cast $B
+  ;; CHECK-NEXT:    (local.get $x)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (call $called
+  ;; CHECK-NEXT:   (ref.cast $B
+  ;; CHECK-NEXT:    (local.get $x)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (call $called
+  ;; CHECK-NEXT:   (ref.cast $C
+  ;; CHECK-NEXT:    (local.get $x)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (call $called
+  ;; CHECK-NEXT:   (ref.cast $C
+  ;; CHECK-NEXT:    (local.get $x)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $caller (export "out") (param $x anyref)
+    ;; Casts of A are refined.
+    (call $called
+      (ref.cast $A
+        (local.get $x)
+      )
+    )
+    ;; This will be refined to B.
+    (call $called
+      (ref.cast null $A
+        (local.get $x)
+      )
+    )
+    ;; Casts of B turn or stay as non-nullable B.
+    (call $called
+      (ref.cast $B
+        (local.get $x)
+      )
+    )
+    (call $called
+      (ref.cast null $B
+        (local.get $x)
+      )
+    )
+    ;; This cast of C remains the same.
+    (call $called
+      (ref.cast $C
+        (local.get $x)
+      )
+    )
+    ;; A nullable C passed into a cast of non-null B means this value must be a
+    ;; non-nullable C.
+    (call $called
+      (ref.cast null $C
+        (local.get $x)
+      )
+    )
+  )
+)
+
+;; A cast of an array.
+(module
+  ;; CHECK:      (type $anyref_=>_none (func (param anyref)))
+
+  ;; CHECK:      (type $A (array (mut i32)))
+  (type $A (array (mut i32)))
+
+  ;; CHECK:      (export "out" (func $caller))
+
+  ;; CHECK:      (func $called (type $anyref_=>_none) (param $x anyref)
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (ref.cast $A
+  ;; CHECK-NEXT:    (local.get $x)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $called (param $x anyref)
+    (drop
+      (ref.cast $A
+        (local.get $x)
+      )
+    )
+  )
+
+  ;; CHECK:      (func $caller (type $anyref_=>_none) (param $x anyref)
+  ;; CHECK-NEXT:  (call $called
+  ;; CHECK-NEXT:   (ref.cast $A
+  ;; CHECK-NEXT:    (local.get $x)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $caller (export "out") (param $x anyref)
+    (call $called
+      ;; This will be refined to a non-nullable cast.
```

Comment:
Should we test with array heap type subtyping as well? Same for functions.

Reply:
There isn't any struct-specific code in the TNHOracle logic, so I was thinking that might be a little redundant. Even the array cast felt a little unneeded to me. But I guess it could be useful for future refactorings perhaps..?

Reply:
Ah, if the code isn't struct-specific, then I'm ok not doing this.